docs: alter streaming jobs (#1660)
risingwavelabs/risingwave-docs: docs/manage/alter-streaming.md

```

The new column will be `NULL` for existing records. 

## Alter a materialized view

To alter a materialized view, you will need to create the new materialized view and drop the old materialized view. 

For example, suppose we want to add a new column to the materialized view `mv1`:
    
```sql
ALTER TABLE customers ADD COLUMN birth_date;
```

Here we create a new materialized view `mv1_new` with the new column `sales_count`:
    
```sql
CREATE MATERIALIZED VIEW mv1_new AS
    SELECT
        customer_id,
        SUM(total_price) AS sales_amount,
        COUNT(*) AS sales_count -- The new column
    FROM test_source
    GROUP BY customer_id;
```

After the new materialized view is created, we can drop the old materialized view `mv1` and rename `mv1_new` to `mv1`:

```sql
DROP SINK mv1;
ALTER SINK mv1_new RENAME TO mv1;
```

## Alter a sink

To alter a sink, you will need to create a new sink and drop the old sink. Please check the example from the last section.

Specifically, if your sink is created based on a materialized view i.e. created with `CREATE SINK ... FROM ...` statement, you may optionally specify `without_backfill = true` to skip the existing data.

```sql

> CREATE SINK ... FROM ... WITH (without_backfill = true).

Author: I think that example won't work. Let's discuss on Slack

> CREATE SINK ... FROM ... WITH (without_backfill = true).

Author: Conclusion: It can't work. So let's delete them now.

Given a table `adult_users` that tracks the number of users aged ≥ 18. 

```sql
CREATE MATERIALIZED VIEW adult_users AS
	SELECT
	  COUNT(*) as user_count
	FROM users
  WHERE age >= 18;
```

Later, it turns out that the legal definition for adulthood should be ≥16. Intuitively, one may attempt to simply modify the filter condition from `age >= 18` to `age >= 16`. However, this is not possible due to the nature of stream processing: records with age between 16 and 18 were already filtered out. Consequently, the only way to refill the missing data is to recompute the entire stream from scratch.

Therefore, as a best practice, it is recommended to persistently store the source data in a long-term storage, e.g, [a RisingWave Table](https://docs.risingwave.com/docs/current/sql-create-table/). This enables recomputing the materialized view when altering logic is necessary.

## Future Roadmap

> We are actively developing features to enhance logic alteration capabilities:

Author: Shall we present this part on documents a bit later? When the function (table sink) is ready. @st1page is working on some testing and bugfixes now. I suppose it can be ready soon.

Author: Removed. Let's add these back when we support these capabilities. ETA: new version

---
id: alter-streaming
title: Alter a streaming job
description: Update an existing streaming job.
slug: /alter-streaming
---
<head>
  <link rel="canonical" href="https://docs.risingwave.com/docs/current/alter-streaming/" />
</head>

> This document covers the mechanisms for altering logic within streaming pipelines in RisingWave. Understanding these mechanisms is crucial for managing and evolving your data processing workflows.

Reviewer: ```suggestion
This document explains how to modify the logic in streaming pipelines within RisingWave. Understanding these mechanisms is essential for effectively managing your data processing workflows.
```

## Alter a table or source

To add or drop columns from a table or source, simply use the [ALTER TABLE](https://docs.risingwave.com/docs/dev/sql-alter-table/) or [ALTER SOURCE](https://docs.risingwave.com/docs/dev/sql-alter-source/) command. For example:

```sql
ALTER TABLE customers ADD COLUMN birth_date;

ALTER SOURCE test_source ADD COLUMN birth_date;
```

The new column will be `NULL` for existing records. 

## Alter a materialized view

> To alter a materialized view, you will need to create the new materialized view and drop the old materialized view. 

Reviewer: ```suggestion
To alter a materialized view, you need to create the new materialized view and drop the old materialized view. 
```

For example, suppose we want to add a new column to the materialized view `mv1`:
    
```sql

> ALTER TABLE customers ADD COLUMN birth_date;

Reviewer: ```suggestion
CREATE MATERIALIZED VIEW mv1 AS
    SELECT
        customer_id,
        SUM(total_price) AS sales_amount,
    FROM test_source
    GROUP BY customer_id;
```

```

Here we create a new materialized view `mv1_new` with the new column `sales_count`:
    
```sql
CREATE MATERIALIZED VIEW mv1_new AS
    SELECT
        customer_id,
        SUM(total_price) AS sales_amount,
        COUNT(*) AS sales_count -- The new column
    FROM test_source
    GROUP BY customer_id;
```

After the new materialized view is created, we can drop the old materialized view `mv1` and rename `mv1_new` to `mv1`:

```sql

> DROP SINK mv1;

Reviewer: ```suggestion
DROP MATERIALIZED VIEW mv1;
```

> ALTER SINK mv1_new RENAME TO mv1;

Reviewer: ```suggestion
ALTER MATERIALIZED VIEW mv1_new RENAME TO mv1;
```

```

## Alter a sink

To alter a sink, you will need to create a new sink and drop the old sink. Please check the example from the last section.

> Specifically, if your sink is created based on a materialized view i.e. created with `CREATE SINK ... FROM ...` statement, you may optionally specify `without_backfill = true` to skip the existing data.

Reviewer: ```suggestion
If your sink is created based on a materialized view using the CREATE SINK ... FROM ... statement, you have the option to specify `without_backfill = true` to exclude existing data.
```

```sql
CREATE SINK ... FROM ... WITH (without_backfill = true).
```

> ## Why I cannot modify a streaming job in place?

Reviewer: ```suggestion
## Why is it not possible to modify a streaming job in place?
```

> Streaming systems like RisingWave need to maintain **internal state** for streaming operators, such as joins and aggregations. Generally, modifying a materialized view would require consistent changes to the internal state accordingly, which is not always feasible. Let’s take the example below to illustrate it.

Reviewer: ```suggestion
Streaming systems like RisingWave need to maintain **internal state** for streaming operators, such as joins and aggregations. Generally, modifying a materialized view would require consistent changes to the internal state accordingly, which is not always feasible. Let’s see an example.
```

Given a table `adult_users` that tracks the number of users aged ≥ 18. 

```sql
CREATE MATERIALIZED VIEW adult_users AS
  SELECT
    COUNT(*) as user_count
  FROM users
  WHERE age >= 18;
```

> Later, it turns out that the legal definition for adulthood should be ≥16. Intuitively, one may attempt to simply modify the filter condition from `age >= 18` to `age >= 16`. However, this is not possible due to the nature of stream processing: records with age between 16 and 18 were already filtered out. Consequently, the only way to refill the missing data is to recompute the entire stream from scratch.

Reviewer: ```suggestion
It was discovered later that the legal definition for adulthood should be set at ≥16. Initially, one might consider modifying the filter condition from `age >= 18` to `age >= 16` as a straightforward solution. However, this is not feasible in stream processing since records with ages between 16 and 18 have already been filtered out. Therefore, the only option to restore the missing data is to recompute the entire stream from the beginning.
```

> Therefore, as a best practice, it is recommended to persistently store the source data in a long-term storage, e.g, [a RisingWave Table](https://docs.risingwave.com/docs/current/sql-create-table/). This enables recomputing the materialized view when altering logic is necessary.

Reviewer: ```suggestion
Therefore, we recommend persistently storing the source data in a long-term storage solution, such as [a RisingWave table](/sql/commands/sql-create-table/). This allows for the recomputation of the materialized view when altering the logic becomes necessary.
```
